src/ripple/rpc/impl/ServerHandlerImp.cpp

```diff
@@ -540,185 +553,194 @@ ServerHandlerImp::processRequest (Port const& port,
-    // Determine role/usage so we can charge for invalid requests
-    Json::Value const& method = jsonRPC [jss::method];
-
-    auto role = Role::FORBID;
-    auto required = RPC::roleRequired(method.asString());
-    if (jsonRPC.isMember(jss::params) &&
-        jsonRPC[jss::params].isArray() &&
-        jsonRPC[jss::params].size() > 0 &&
-        jsonRPC[jss::params][Json::UInt(0)].isObject())
-    {
-        role = requestRole(required, port, jsonRPC[jss::params][Json::UInt(0)],
-            remoteIPAddress, user);
-    }
-    else
+    auto size = jsonOrig.isArray() ? jsonOrig.size() : 1;
+    for (unsigned i = 0; i < size; ++i)
     {
-        role = requestRole(required, port, Json::objectValue,
-            remoteIPAddress, user);
-    }
+        Json::Value const& jsonRPC = jsonOrig.isArray() ? jsonOrig[i] : jsonOrig;
+        /* ---------------------------------------------------------------------- */
+        // Determine role/usage so we can charge for invalid requests
+        Json::Value const& method = jsonRPC [jss::method];
 
-    Resource::Consumer usage;
-    if (isUnlimited(role))
-    {
-        usage = m_resourceManager.newUnlimitedEndpoint(
-            remoteIPAddress.to_string());
-    }
-    else
-    {
-        usage = m_resourceManager.newInboundEndpoint(remoteIPAddress);
-        if (usage.disconnect())
+        auto role = Role::FORBID;
+        auto required = RPC::roleRequired(method.asString());
+        if (jsonRPC.isMember(jss::params) &&
+            jsonRPC[jss::params].isArray() &&
+            jsonRPC[jss::params].size() > 0 &&
+            jsonRPC[jss::params][Json::UInt(0)].isObject())
         {
-            HTTPReply(503, "Server is overloaded", output, rpcJ);
-            return;
+            role = requestRole(required, port, jsonRPC[jss::params][Json::UInt(0)],
+                remoteIPAddress, user);
+        }
+        else
+        {
+            role = requestRole(required, port, Json::Value{Json::objectValue},
+                remoteIPAddress, user);
         }
-    }
 
-    if (role == Role::FORBID)
-    {
-        usage.charge(Resource::feeInvalidRPC);
-        HTTPReply (403, "Forbidden", output, rpcJ);
-        return;
-    }
+        Resource::Consumer usage;
+        if (isUnlimited(role))
+        {
+            usage = m_resourceManager.newUnlimitedEndpoint(
+                remoteIPAddress.to_string());
+        }
+        else
+        {
+            usage = m_resourceManager.newInboundEndpoint(remoteIPAddress);
+            if (usage.disconnect())
+            {
+                HTTPReply(503, "Server is overloaded", output, rpcJ);
+                return;
+            }
+        }
 
-    if (! method)
-    {
-        usage.charge(Resource::feeInvalidRPC);
-        HTTPReply (400, "Null method", output, rpcJ);
-        return;
-    }
+        if (role == Role::FORBID)
+        {
+            usage.charge(Resource::feeInvalidRPC);
+            HTTPReply (403, "Forbidden", output, rpcJ);
```

Comment:
Ok, I'll start working that direction, as it would be a required precondition for doing the batch work anyway.  But I'm little concerned that changing behavior will cause json client breakage (non-rippled clients).  What are your thoughts about that?